File: coding_py/bj2239.py
sudoku = []
row = [[False for _ in range(9)] for _ in range(9)]
col = [[False for _ in range(9)] for _ in range(9)]
sqare = [[False for _ in range(9)] for _ in range(9)]
haveto = []

for i in range(9):
    sudoku.append(list(map(int, list(input().rstrip()))))
    for j in range(9):
        num = sudoku[i][j]-1
        if num != -1:
            row[i][num] = True
            col[j][num] = True
            sqare[i//3*3 + j//3][num] = True
        else:
            haveto.append([i, j])


def dfs(idx):
    if idx == len(haveto):
        for line in sudoku:
            print(''.join(map(str, line)))
        exit()

    x, y = haveto[idx][0], haveto[idx][1]

    for i in range(9):
        if row[x][i] or col[y][i] or sqare[x//3*3 + y//3][i]:
            continue
        row[x][i] = col[y][i] = sqare[x//3*3+y//3][i] = True
        sudoku[x][y] = i+1
        dfs(idx+1)
        row[x][i] = col[y][i] = sqare[x//3*3+y//3][i] = False


dfs(0)

File: algo_cpp/bj2239.cpp
#include <vector>
#include <cstdlib>
#include <cstdio>
using namespace std;

vector<pair<int, int>> f;
int cnt;
bool row[9][9];
bool col[9][9];
bool sqr[9][9];
int sudoku[9][9];

void dfs(int idx) {
	if (idx == cnt) {
		for (int i = 0; i < 9; i++) {
			for (int j = 0; j < 9; j++) {
				printf("%d",sudoku[i][j]);
			}
			printf("\n");
		}
		exit(0);
	}
	int x, y;
	x = f[idx].first;
	y = f[idx].second;
	for (int i = 0; i < 9; i++) {
		if (!row[x][i] && !col[y][i] && !sqr[x / 3 * 3 + y / 3][i]) {
			row[x][i] = col[y][i] = sqr[x / 3 * 3 + y / 3][i] = 1;
			sudoku[x][y] = i + 1;
			dfs(idx + 1);
			sudoku[x][y] = row[x][i] = col[y][i] = sqr[x / 3 * 3 + y / 3][i] = 0;
		}
	}
}

int main(void) {
	for (int i = 0; i < 9; i++) {
		for (int j = 0; j < 9; j++) {
			sudoku[i][j] = getchar()-'0';
			if (!sudoku[i][j])
				f.push_back({ i,j });
			else
			{
				row[i][sudoku[i][j] - 1] = 1;
				col[j][sudoku[i][j] - 1] = 1;
				sqr[i/3*3+j/3][sudoku[i][j] - 1] = 1;
			}
		}
		getchar();
	}

	cnt = f.size();
	dfs(0);
	return 0;
}